File: tools/apigen/testy.rb
require 'ffi'

class Vector2 < FFI::Struct
  layout \
    x: :float,
    y: :float
end

class Address < FFI::Struct
  layout \
    state: :int64,
    zip: :int64
end

class Contact < FFI::Struct
  layout \
    first_name: :int64,
    last_name: :int64,
    address: object
  state
  zip
.by_value
end

class Fod < FFI::Struct
  layout ref: :uint64

  def name
    return DLL.fidget_fod_get_name(self)
  end
  def name=(v)
    DLL.fidget_fod_set_name(self, v)
  end

  def count
    return DLL.fidget_fod_get_count(self)
  end
  def count=(v)
    DLL.fidget_fod_set_count(self, v)
  end
end

class Boz < FFI::Struct
  layout ref: :uint64

  def name
    return DLL.fidget_boz_get_name(self)
  end
  def name=(v)
    DLL.fidget_boz_set_name(self, v)
  end

  def fod
    return DLL.fidget_boz_get_fod(self)
  end
  def fod=(v)
    DLL.fidget_boz_set_fod(self, v)
  end
end

AlignSomething = :uint64
AS_DEFAULT = 0
AS_TOP = 1
AS_BOTTOM = 2
AS_RIGHT = 3
AS_LEFT = 4

TextCase = :uint64
TC_NORMAL = 0
TC_UPPER = 1
TC_LOWER = 2
TC_TITLE = 3

PaintKind = :uint64
PK_SOLID = 0
PK_IMAGE = 1
PK_IMAGE_TILED = 2
PK_GRADIENT_LINEAR = 3
PK_GRADIENT_RADIAL = 4
PK_GRADIENT_ANGULAR = 5

BlendMode = :uint64
BM_NORMAL = 0
BM_DARKEN = 1
BM_MULTIPLY = 2
BM_COLOR_BURN = 3
BM_LIGHTEN = 4
BM_SCREEN = 5
BM_COLOR_DODGE = 6
BM_OVERLAY = 7
BM_SOFT_LIGHT = 8
BM_HARD_LIGHT = 9
BM_DIFFERENCE = 10
BM_EXCLUSION = 11
BM_HUE = 12
BM_SATURATION = 13
BM_COLOR = 14
BM_LUMINOSITY = 15
BM_MASK = 16
BM_OVERWRITE = 17
BM_SUBTRACT_MASK = 18
BM_EXCLUDE_MASK = 19

class ColorRGBX < FFI::Struct
  layout \
    r: :uint8,
    g: :uint8,
    b: :uint8,
    a: :uint8
end

class ColorStop2 < FFI::Struct
  layout \
    color: object
  r
  g
  b
  a
.by_value,
    position: :float
end

class Paint2 < FFI::Struct
  layout \
    kind: enum
  pkSolid, pkImage, pkImageTiled, pkGradientLinear, pkGradientRadial,
  pkGradientAngular,
    blend_mode: enum
  bmNormal, bmDarken, bmMultiply, bmColorBurn, bmLighten, bmScreen,
  bmColorDodge, bmOverlay, bmSoftLight, bmHardLight, bmDifference, bmExclusion,
  bmHue, bmSaturation, bmColor, bmLuminosity, bmMask, bmOverwrite,
  bmSubtractMask, bmExcludeMask,
    gradient_stops: ColorStop2.by_value
end

class Typeface2 < FFI::Struct
  layout ref: :uint64

  def file_path
    return DLL.fidget_typeface2_get_file_path(self)
  end
  def file_path=(v)
    DLL.fidget_typeface2_set_file_path(self, v)
  end
end

class Font2 < FFI::Struct
  layout ref: :uint64

  def typeface
    return DLL.fidget_font2_get_typeface(self)
  end
  def typeface=(v)
    DLL.fidget_font2_set_typeface(self, v)
  end

  def size
    return DLL.fidget_font2_get_size(self)
  end
  def size=(v)
    DLL.fidget_font2_set_size(self, v)
  end

  def line_height
    return DLL.fidget_font2_get_line_height(self)
  end
  def line_height=(v)
    DLL.fidget_font2_set_line_height(self, v)
  end

  def paint
    return DLL.fidget_font2_get_paint(self)
  end
  def paint=(v)
    DLL.fidget_font2_set_paint(self, v)
  end

  def text_case
    return DLL.fidget_font2_get_text_case(self)
  end
  def text_case=(v)
    DLL.fidget_font2_set_text_case(self, v)
  end

  def underline
    return DLL.fidget_font2_get_underline(self)
  end
  def underline=(v)
    DLL.fidget_font2_set_underline(self, v)
  end

  def strikethrough
    return DLL.fidget_font2_get_strikethrough(self)
  end
  def strikethrough=(v)
    DLL.fidget_font2_set_strikethrough(self, v)
  end

  def no_kerning_adjustments
    return DLL.fidget_font2_get_no_kerning_adjustments(self)
  end
  def no_kerning_adjustments=(v)
    DLL.fidget_font2_set_no_kerning_adjustments(self, v)
  end
end


module DLL
  extend FFI::Library
  ffi_lib '/Users/me/p/fidget2/tools/apigen/libfidget.dylib'
  callback :fidget_cb, [], :void

  attach_function :fidget_input_code, [ :int64, :int64, :int64, :int64 ], :bool
  attach_function :fidget_test_numbers, [ :int8, :uint8, :int16, :uint16, :int32, :uint32, :int64, :uint64, :int64, :uint64, :float, :double, :double ], :bool
  attach_function :fidget_call_me_maybe, [ :string ], :void
  attach_function :fidget_flight_club_rule, [ :int64 ], :string
  attach_function :fidget_cat_str, [ :string, :string, :string, :string, :string ], :string
  attach_function :fidget_give_vec, [ Vector2.by_value ], :void
  attach_function :fidget_take_vec, [  ], Vector2.by_value
  attach_function :fidget_take_contact, [  ], Contact.by_value
  attach_function :fidget_fod_get_name, [Fod.by_value], :string
  attach_function :fidget_fod_set_name, [Fod.by_value, :string], :void
  attach_function :fidget_fod_get_count, [Fod.by_value], :int64
  attach_function :fidget_fod_set_count, [Fod.by_value, :int64], :void
  attach_function :fidget_take_fod, [  ], Fod.by_value
  attach_function :fidget_boz_get_name, [Boz.by_value], :string
  attach_function :fidget_boz_set_name, [Boz.by_value, :string], :void
  attach_function :fidget_boz_get_fod, [Boz.by_value], Fod.by_value
  attach_function :fidget_boz_set_fod, [Boz.by_value, Fod.by_value], :void
  attach_function :fidget_take_boz, [  ], Boz.by_value
  attach_function :fidget_repeat_enum, [ AlignSomething ], AlignSomething
  attach_function :fidget_call_me_back, [ :fidget_cb ], :void
  attach_function :fidget_take_seq, [  ], uint64.by_value
  attach_function :fidget_give_seq, [ uint64.by_value ], :void
  attach_function :fidget_give_seq_of_vector2, [ Vector2.by_value ], :void
  attach_function :fidget_take_seq_of_vector2, [  ], Vector2.by_value
  attach_function :fidget_give_seq_of_boz, [ Boz.by_value ], :void
  attach_function :fidget_take_seq_of_boz, [  ], Boz.by_value
  attach_function :fidget_typeface2_get_file_path, [Typeface2.by_value], :string
  attach_function :fidget_typeface2_set_file_path, [Typeface2.by_value, :string], :void
  attach_function :fidget_font2_get_typeface, [Font2.by_value], Typeface2.by_value
  attach_function :fidget_font2_set_typeface, [Font2.by_value, Typeface2.by_value], :void
  attach_function :fidget_font2_get_size, [Font2.by_value], :float
  attach_function :fidget_font2_set_size, [Font2.by_value, :float], :void
  attach_function :fidget_font2_get_line_height, [Font2.by_value], :float
  attach_function :fidget_font2_set_line_height, [Font2.by_value, :float], :void
  attach_function :fidget_font2_get_paint, [Font2.by_value], object
  kind
  blendMode
  gradientStops
.by_value
  attach_function :fidget_font2_set_paint, [Font2.by_value, object
  kind
  blendMode
  gradientStops
.by_value], :void
  attach_function :fidget_font2_get_text_case, [Font2.by_value], enum
  tcNormal, tcUpper, tcLower, tcTitle
  attach_function :fidget_font2_set_text_case, [Font2.by_value, enum
  tcNormal, tcUpper, tcLower, tcTitle], :void
  attach_function :fidget_font2_get_underline, [Font2.by_value], :bool
  attach_function :fidget_font2_set_underline, [Font2.by_value, :bool], :void
  attach_function :fidget_font2_get_strikethrough, [Font2.by_value], :bool
  attach_function :fidget_font2_set_strikethrough, [Font2.by_value, :bool], :void
  attach_function :fidget_font2_get_no_kerning_adjustments, [Font2.by_value], :bool
  attach_function :fidget_font2_set_no_kerning_adjustments, [Font2.by_value, :bool], :void
  attach_function :fidget_read_font2, [ :string ], Font2.by_value
end
def input_code(a, b, c, d)
  return DLL.fidget_input_code(a, b, c, d)
end

def test_numbers(a, b, c, d, e, f, g, h, i, j, k, l, m)
  return DLL.fidget_test_numbers(a, b, c, d, e, f, g, h, i, j, k, l, m)
end

def call_me_maybe(phone)
  return DLL.fidget_call_me_maybe(phone)
end

def flight_club_rule(n)
  return DLL.fidget_flight_club_rule(n)
end

def cat_str(a, b, c, d, e)
  return DLL.fidget_cat_str(a, b, c, d, e)
end

def give_vec(v)
  return DLL.fidget_give_vec(v)
end

def take_vec()
  return DLL.fidget_take_vec()
end

def take_contact()
  return DLL.fidget_take_contact()
end

def take_fod()
  return DLL.fidget_take_fod()
end

def take_boz()
  return DLL.fidget_take_boz()
end

def repeat_enum(e)
  return DLL.fidget_repeat_enum(e)
end

def call_me_back(cb)
  return DLL.fidget_call_me_back(cb)
end

def take_seq()
  return DLL.fidget_take_seq()
end

def give_seq(s)
  return DLL.fidget_give_seq(s)
end

def give_seq_of_vector2(s)
  return DLL.fidget_give_seq_of_vector2(s)
end

def take_seq_of_vector2()
  return DLL.fidget_take_seq_of_vector2()
end

def give_seq_of_boz(s)
  return DLL.fidget_give_seq_of_boz(s)
end

def take_seq_of_boz()
  return DLL.fidget_take_seq_of_boz()
end

def read_font2(font_path)
  return DLL.fidget_read_font2(font_path)
end

File: tools/apigen/testy.js
var ffi = require('ffi-napi');
var Struct = require("ref-struct-napi");

exports.cb = function(f){return ffi.Callback('void', [], f)};

var Vector2 = Struct({
  'x': 'float',
  'y': 'float'});
var Address = Struct({
  'state': 'int64',
  'zip': 'int64'});
var Contact = Struct({
  'firstName': 'int64',
  'lastName': 'int64',
  'address': object
  state
  zip
});
var Fod = Struct({'nimRef': 'pointer'});
var Boz = Struct({'nimRef': 'pointer'});
AlignSomething = 'int64'
exports.AS_DEFAULT = 0
exports.AS_TOP = 1
exports.AS_BOTTOM = 2
exports.AS_RIGHT = 3
exports.AS_LEFT = 4

TextCase = 'int64'
exports.TC_NORMAL = 0
exports.TC_UPPER = 1
exports.TC_LOWER = 2
exports.TC_TITLE = 3

PaintKind = 'int64'
exports.PK_SOLID = 0
exports.PK_IMAGE = 1
exports.PK_IMAGE_TILED = 2
exports.PK_GRADIENT_LINEAR = 3
exports.PK_GRADIENT_RADIAL = 4
exports.PK_GRADIENT_ANGULAR = 5

BlendMode = 'int64'
exports.BM_NORMAL = 0
exports.BM_DARKEN = 1
exports.BM_MULTIPLY = 2
exports.BM_COLOR_BURN = 3
exports.BM_LIGHTEN = 4
exports.BM_SCREEN = 5
exports.BM_COLOR_DODGE = 6
exports.BM_OVERLAY = 7
exports.BM_SOFT_LIGHT = 8
exports.BM_HARD_LIGHT = 9
exports.BM_DIFFERENCE = 10
exports.BM_EXCLUSION = 11
exports.BM_HUE = 12
exports.BM_SATURATION = 13
exports.BM_COLOR = 14
exports.BM_LUMINOSITY = 15
exports.BM_MASK = 16
exports.BM_OVERWRITE = 17
exports.BM_SUBTRACT_MASK = 18
exports.BM_EXCLUDE_MASK = 19

var ColorRGBX = Struct({
  'r': 'uint8',
  'g': 'uint8',
  'b': 'uint8',
  'a': 'uint8'});
var ColorStop2 = Struct({
  'color': object
  r
  g
  b
  a
,
  'position': 'float'});
var Paint2 = Struct({
  'kind': enum
  pkSolid, pkImage, pkImageTiled, pkGradientLinear, pkGradientRadial,
  pkGradientAngular,
  'blendMode': enum
  bmNormal, bmDarken, bmMultiply, bmColorBurn, bmLighten, bmScreen,
  bmColorDodge, bmOverlay, bmSoftLight, bmHardLight, bmDifference, bmExclusion,
  bmHue, bmSaturation, bmColor, bmLuminosity, bmMask, bmOverwrite,
  bmSubtractMask, bmExcludeMask,
  'gradientStops': ColorStop2});
var Typeface2 = Struct({'nimRef': 'pointer'});
var Font2 = Struct({'nimRef': 'pointer'});

var dllPath = ""
if(process.platform == "win32") {
  dllPath = 'fidget.dll'
} else if (process.platform == "darwin") {
  dllPath = __dirname + '/libfidget.dylib'
} else {
  dllPath = __dirname + '/libfidget.so'
}

var fidget = ffi.Library(dllPath, {
  'fidget_input_code': ['bool', ['int64', 'int64', 'int64', 'int64']],
  'fidget_test_numbers': ['bool', ['int8', 'uint8', 'int16', 'uint16', 'int32', 'uint32', 'int64', 'uint64', 'int64', 'uint64', 'float', 'double', 'double']],
  'fidget_call_me_maybe': ['void', ['string']],
  'fidget_flight_club_rule': ['string', ['int64']],
  'fidget_cat_str': ['string', ['string', 'string', 'string', 'string', 'string']],
  'fidget_give_vec': ['void', [Vector2]],
  'fidget_take_vec': [Vector2, []],
  'fidget_take_contact': [Contact, []],
  'fidget_fod_get_name': ['string', [Fod]],
  'fidget_fod_set_name': ['void', [Fod, 'string']],
  'fidget_fod_get_count': ['int64', [Fod]],
  'fidget_fod_set_count': ['void', [Fod, 'int64']],
  'fidget_take_fod': [Fod, []],
  'fidget_boz_get_name': ['string', [Boz]],
  'fidget_boz_set_name': ['void', [Boz, 'string']],
  'fidget_boz_get_fod': [Fod, [Boz]],
  'fidget_boz_set_fod': ['void', [Boz, Fod]],
  'fidget_take_boz': [Boz, []],
  'fidget_repeat_enum': [AlignSomething, [AlignSomething]],
  'fidget_call_me_back': ['void', ['pointer']],
  'fidget_take_seq': [uint64, []],
  'fidget_give_seq': ['void', [uint64]],
  'fidget_give_seq_of_vector2': ['void', [Vector2]],
  'fidget_take_seq_of_vector2': [Vector2, []],
  'fidget_give_seq_of_boz': ['void', [Boz]],
  'fidget_take_seq_of_boz': [Boz, []],
  'fidget_typeface2_get_file_path': ['string', [Typeface2]],
  'fidget_typeface2_set_file_path': ['void', [Typeface2, 'string']],
  'fidget_font2_get_typeface': [Typeface2, [Font2]],
  'fidget_font2_set_typeface': ['void', [Font2, Typeface2]],
  'fidget_font2_get_size': ['float', [Font2]],
  'fidget_font2_set_size': ['void', [Font2, 'float']],
  'fidget_font2_get_line_height': ['float', [Font2]],
  'fidget_font2_set_line_height': ['void', [Font2, 'float']],
  'fidget_font2_get_paint': [object
  kind
  blendMode
  gradientStops
, [Font2]],
  'fidget_font2_set_paint': ['void', [Font2, object
  kind
  blendMode
  gradientStops
]],
  'fidget_font2_get_text_case': [enum
  tcNormal, tcUpper, tcLower, tcTitle, [Font2]],
  'fidget_font2_set_text_case': ['void', [Font2, enum
  tcNormal, tcUpper, tcLower, tcTitle]],
  'fidget_font2_get_underline': ['bool', [Font2]],
  'fidget_font2_set_underline': ['void', [Font2, 'bool']],
  'fidget_font2_get_strikethrough': ['bool', [Font2]],
  'fidget_font2_set_strikethrough': ['void', [Font2, 'bool']],
  'fidget_font2_get_no_kerning_adjustments': ['bool', [Font2]],
  'fidget_font2_set_no_kerning_adjustments': ['void', [Font2, 'bool']],
  'fidget_read_font2': [Font2, ['string']],
});

exports.inputCode = fidget.fidget_input_code;
exports.testNumbers = fidget.fidget_test_numbers;
exports.callMeMaybe = fidget.fidget_call_me_maybe;
exports.flightClubRule = fidget.fidget_flight_club_rule;
exports.catStr = fidget.fidget_cat_str;
exports.Vector2 = Vector2;
exports.giveVec = fidget.fidget_give_vec;
exports.takeVec = fidget.fidget_take_vec;
exports.Address = Address;
exports.Contact = Contact;
exports.takeContact = fidget.fidget_take_contact;
exports.Fod = Fod;
Object.defineProperty(Fod.prototype, 'name', {
  get: function() {return fidget.fidget_fod_get_name(this)},
  set: function(v) {fidget.fidget_fod_set_name(this, v)}
});
Object.defineProperty(Fod.prototype, 'count', {
  get: function() {return fidget.fidget_fod_get_count(this)},
  set: function(v) {fidget.fidget_fod_set_count(this, v)}
});
exports.takeFod = fidget.fidget_take_fod;
exports.Boz = Boz;
Object.defineProperty(Boz.prototype, 'name', {
  get: function() {return fidget.fidget_boz_get_name(this)},
  set: function(v) {fidget.fidget_boz_set_name(this, v)}
});
Object.defineProperty(Boz.prototype, 'fod', {
  get: function() {return fidget.fidget_boz_get_fod(this)},
  set: function(v) {fidget.fidget_boz_set_fod(this, v)}
});
exports.takeBoz = fidget.fidget_take_boz;
exports.repeatEnum = fidget.fidget_repeat_enum;
exports.callMeBack = fidget.fidget_call_me_back;
exports.takeSeq = fidget.fidget_take_seq;
exports.giveSeq = fidget.fidget_give_seq;
exports.giveSeqOfVector2 = fidget.fidget_give_seq_of_vector2;
exports.takeSeqOfVector2 = fidget.fidget_take_seq_of_vector2;
exports.giveSeqOfBoz = fidget.fidget_give_seq_of_boz;
exports.takeSeqOfBoz = fidget.fidget_take_seq_of_boz;
exports.ColorRGBX = ColorRGBX;
exports.ColorStop2 = ColorStop2;
exports.Paint2 = Paint2;
exports.Typeface2 = Typeface2;
Object.defineProperty(Typeface2.prototype, 'filePath', {
  get: function() {return fidget.fidget_typeface2_get_file_path(this)},
  set: function(v) {fidget.fidget_typeface2_set_file_path(this, v)}
});
exports.Font2 = Font2;
Object.defineProperty(Font2.prototype, 'typeface', {
  get: function() {return fidget.fidget_font2_get_typeface(this)},
  set: function(v) {fidget.fidget_font2_set_typeface(this, v)}
});
Object.defineProperty(Font2.prototype, 'size', {
  get: function() {return fidget.fidget_font2_get_size(this)},
  set: function(v) {fidget.fidget_font2_set_size(this, v)}
});
Object.defineProperty(Font2.prototype, 'lineHeight', {
  get: function() {return fidget.fidget_font2_get_line_height(this)},
  set: function(v) {fidget.fidget_font2_set_line_height(this, v)}
});
Object.defineProperty(Font2.prototype, 'paint', {
  get: function() {return fidget.fidget_font2_get_paint(this)},
  set: function(v) {fidget.fidget_font2_set_paint(this, v)}
});
Object.defineProperty(Font2.prototype, 'textCase', {
  get: function() {return fidget.fidget_font2_get_text_case(this)},
  set: function(v) {fidget.fidget_font2_set_text_case(this, v)}
});
Object.defineProperty(Font2.prototype, 'underline', {
  get: function() {return fidget.fidget_font2_get_underline(this)},
  set: function(v) {fidget.fidget_font2_set_underline(this, v)}
});
Object.defineProperty(Font2.prototype, 'strikethrough', {
  get: function() {return fidget.fidget_font2_get_strikethrough(this)},
  set: function(v) {fidget.fidget_font2_set_strikethrough(this, v)}
});
Object.defineProperty(Font2.prototype, 'noKerningAdjustments', {
  get: function() {return fidget.fidget_font2_get_no_kerning_adjustments(this)},
  set: function(v) {fidget.fidget_font2_set_no_kerning_adjustments(this, v)}
});
exports.readFont2 = fidget.fidget_read_font2;
